idx/cassandra/cassandra.go

```diff
@@ -366,10 +384,177 @@ func (c *CasIdx) rebuildIndex() {
 			atomic.AddUint32(&num, uint32(c.MemoryIndex.LoadPartition(p, defs)))
 		}(partition)
 	}
+
+	if memory.MetaTagSupport {
+		wg.Add(1)
+		go func() {
+			gate <- struct{}{}
+
+			records := c.loadMetaRecords()
+			for orgId, values := range records {
+				c.applyMetaRecords(orgId, values)
+			}
+
+			wg.Done()
+			<-gate
+		}()
+	}
+
 	wg.Wait()
 	log.Infof("cassandra-idx: Rebuilding Memory Index Complete. Imported %d. Took %s", num, time.Since(pre))
 }
 
+func (c *CasIdx) loadMetaRecords() map[uint32][]tagquery.MetaTagRecord {
+	q := fmt.Sprintf("SELECT orgid, expressions, metatags FROM %s", c.Config.MetaRecordTable)
+	var orgId uint32
+	var expressions, metatags string
+	iter := c.Session.Query(q).Iter()
+
+	res := make(map[uint32][]tagquery.MetaTagRecord)
+	for iter.Scan(&orgId, &expressions, &metatags) {
+		record := tagquery.MetaTagRecord{}
+		err := json.Unmarshal([]byte(expressions), &record.Expressions)
+		if err != nil {
+			log.Errorf("cassandra-idx: LoadMetaRecords() could not parse stored expressions (%s): %s", expressions, err)
+			continue
+		}
+
+		err = json.Unmarshal([]byte(metatags), &record.MetaTags)
+		if err != nil {
+			log.Errorf("cassandra-idx: LoadMetaRecords() could not parse stored metatags (%s): %s", metatags, err)
+			continue
+		}
+
+		res[orgId] = append(res[orgId], record)
+	}
+
+	return res
+}
+
+func (c *CasIdx) applyMetaRecords(orgId uint32, records []tagquery.MetaTagRecord) {
+	_, _, err := c.MemoryIndex.MetaTagRecordSwap(orgId, records, false)
+	if err != nil {
+		log.Errorf("cassandra-idx: applyMetaRecords() failed to apply meta record: %s", err)
+	}
+}
+
+func (c *CasIdx) MetaTagRecordUpsert(orgId uint32, upsertRecord tagquery.MetaTagRecord, persist bool) (tagquery.MetaTagRecord, bool, error) {
+	record, created, err := c.MemoryIndex.MetaTagRecordUpsert(orgId, upsertRecord, persist)
+	if err != nil {
+		return record, created, err
+	}
+
+	if c.Config.updateCassIdx && persist {
+		var err error
+
+		// if a record has no meta tags associated with it, then we delete it
+		if len(record.MetaTags) > 0 {
+			err = c.persistMetaRecord(orgId, record, created)
+		} else {
+			err = c.deleteMetaRecord(orgId, record)
+		}
+
+		if err != nil {
+			log.Errorf("Failed to update meta records in cassandra: %s", err)
+			return record, created, fmt.Errorf("Failed to update cassandra: %s", err)
+		}
+	}
+
+	return record, created, nil
+}
+
+func (c *CasIdx) MetaTagRecordSwap(orgId uint32, records []tagquery.MetaTagRecord, persist bool) (uint32, uint32, error) {
+	added, deleted, err := c.MemoryIndex.MetaTagRecordSwap(orgId, records, persist)
```

Comment:
doesnt adding to the memoryIdx before adding to cassandra lead to things getting out of sync?  eg the memoryIdx gets updated, but the batch update to cassandra fails.

Reply:
the plan is to reload from cassandra at some interval anyway, at this point the state of the memory index would then get reset to what it previously was.
the reason why the updating of the memory index should be done first is mainly because that way we will be able to detect whether anything has changed, or if the posted set of meta records is just the same as the old one. this is implemented in this subsequent PR:
https://github.com/grafana/metrictank/pull/1480/files#diff-c6ce4629577f4e064a1d2f636acc0568R496

Reply:
I guess it would also be reasonable to just always flush to cassandra first, even if there was no change. And if the query to cassandra failed we don't update the memory index. I'm going to do that.

Reply:
I forgot, there is one more reason to do the memory index update first. On upsert we want to be able to know whether a record has been created, or if it only updates an existing one. When one gets created then we set the `createdat` timestamp in cassandra, otherwise we only update the `lastupdate` timestamp.

Reply:
done: https://github.com/grafana/metrictank/pull/1471/commits/6a73386abc2d05c664ea592c64e84b8f548bce46